File: scripts/parse-wordpress-xml.mjs
#!/usr/bin/env node
/**
 * Parse WordPress WXR XML export and extract published posts.
 * No external dependencies — uses regex parsing on well-structured WXR.
 *
 * Usage: node scripts/parse-wordpress-xml.mjs
 */

import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { dirname } from 'path';

const XML_PATH = '/Users/weizy0219/Documents/repos/nanoclaw/groups/telegram_main/blog_backup/WordPress.2026-03-17.xml';
const OUT_PATH = '/Users/weizy0219/repos/johngai-blog/content/weizhiyong-articles.json';

// ── Category mapping ────────────────────────────────────────────────
const CATEGORY_MAP = {
  '开发运维': 'engineering',
  '编程环境': 'engineering',
  'JavaScript': 'engineering',
  'C_CPP_CSharp': 'engineering',
  '其他编程语言': 'engineering',
  '操作系统': 'engineering',
  'LabVIEW': 'engineering',
  '数据库': 'engineering',
  '办公软件': 'engineering',
  '应用软件': 'engineering',
  '工业互联网': 'industry',
  '工业自动化': 'industry',
  '能源技术': 'industry',
  '试验检测': 'industry',
  '工业机器人': 'industry',
  '文学艺术': 'books',
  '影视音乐': 'books',
  '浮生掠影': 'life',
  '经营管理': 'startup',
  '未分类': 'engineering',   // fallback
  '他山之石': 'engineering',  // quotes/references — tech-oriented
};

// ── Helpers ─────────────────────────────────────────────────────────

/** Extract text from CDATA or plain tag content */
function cdata(str) {
  if (!str) return '';
  const m = str.match(/<!\[CDATA\[([\s\S]*?)\]\]>/);
  return m ? m[1] : str.replace(/<[^>]+>/g, '').trim();
}

/** Get first match of a tag within a block */
function tagContent(block, tag) {
  // Handle namespaced tags like wp:post_type, content:encoded
  const escaped = tag.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const re = new RegExp(`<${escaped}[^>]*>([\\s\\S]*?)</${escaped}>`, 'm');
  const m = block.match(re);
  return m ? m[1] : '';
}

/** Convert HTML content to simplified Markdown */
function htmlToMarkdown(html) {
  if (!html) return '';

  let md = html;

  // Remove WordPress block comments
  md = md.replace(/<!--[\s\S]*?-->/g, '');

  // Convert headings
  md = md.replace(/<h1[^>]*>([\s\S]*?)<\/h1>/gi, (_, c) => `# ${stripTags(c).trim()}\n\n`);
  md = md.replace(/<h2[^>]*>([\s\S]*?)<\/h2>/gi, (_, c) => `## ${stripTags(c).trim()}\n\n`);
  md = md.replace(/<h3[^>]*>([\s\S]*?)<\/h3>/gi, (_, c) => `### ${stripTags(c).trim()}\n\n`);
  md = md.replace(/<h4[^>]*>([\s\S]*?)<\/h4>/gi, (_, c) => `#### ${stripTags(c).trim()}\n\n`);

  // Convert images — keep as markdown image
  md = md.replace(/<img[^>]*src="([^"]*)"[^>]*alt="([^"]*)"[^>]*\/?>/gi, (_, src, alt) => `![${alt}](${src})\n\n`);
  md = md.replace(/<img[^>]*src="([^"]*)"[^>]*\/?>/gi, (_, src) => `![](${src})\n\n`);

  // Convert links
  md = md.replace(/<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, (_, href, text) => `[${stripTags(text).trim()}](${href})`);

  // Convert lists
  md = md.replace(/<ul[^>]*>([\s\S]*?)<\/ul>/gi, (_, inner) => {
    return inner.replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, (_, li) => `- ${stripTags(li).trim()}\n`) + '\n';
  });
  md = md.replace(/<ol[^>]*>([\s\S]*?)<\/ol>/gi, (_, inner) => {
    let i = 0;
    return inner.replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, (_, li) => `${++i}. ${stripTags(li).trim()}\n`) + '\n';
  });

  // Convert blockquotes
  md = md.replace(/<blockquote[^>]*>([\s\S]*?)<\/blockquote>/gi, (_, c) => {
    return stripTags(c).trim().split('\n').map(l => `> ${l}`).join('\n') + '\n\n';
  });

  // Convert code blocks
  md = md.replace(/<pre[^>]*><code[^>]*>([\s\S]*?)<\/code><\/pre>/gi, (_, c) => `\`\`\`\n${decodeEntities(c).trim()}\n\`\`\`\n\n`);
  md = md.replace(/<pre[^>]*>([\s\S]*?)<\/pre>/gi, (_, c) => `\`\`\`\n${decodeEntities(stripTags(c)).trim()}\n\`\`\`\n\n`);
  md = md.replace(/<code[^>]*>([\s\S]*?)<\/code>/gi, (_, c) => `\`${decodeEntities(c).trim()}\``);

  // Convert bold/italic/strong/em
  md = md.replace(/<(strong|b)>([\s\S]*?)<\/\1>/gi, (_, __, c) => `**${c.trim()}**`);
  md = md.replace(/<(em|i)>([\s\S]*?)<\/\1>/gi, (_, __, c) => `*${c.trim()}*`);

  // Convert <br> to newline
  md = md.replace(/<br\s*\/?>/gi, '\n');

  // Convert paragraphs
  md = md.replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, (_, c) => `${c.trim()}\n\n`);

  // Remove figure/figcaption wrappers
  md = md.replace(/<\/?figure[^>]*>/gi, '');
  md = md.replace(/<figcaption[^>]*>([\s\S]*?)<\/figcaption>/gi, '');

  // Remove remaining HTML tags
  md = md.replace(/<[^>]+>/g, '');

  // Decode HTML entities
  md = decodeEntities(md);

  // Clean up whitespace: collapse 3+ newlines to 2
  md = md.replace(/\n{3,}/g, '\n\n').trim();

  return md;
}

function stripTags(html) {
  return html.replace(/<[^>]+>/g, '');
}

function decodeEntities(str) {
  return str
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#039;/g, "'")
    .replace(/&nbsp;/g, ' ')
    .replace(/&#(\d+);/g, (_, n) => String.fromCharCode(Number(n)));
}

/** Generate URL-friendly slug from title */
function generateSlug(title) {
  // If title is mostly ASCII/English, just slugify directly
  const ascii = title.replace(/[^\x00-\x7F]/g, '');
  if (ascii.length > title.length * 0.5) {
    return ascii
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/^-|-$/g, '')
      .substring(0, 80);
  }

  // For Chinese titles: use pinyin-like transliteration via simple mapping
  // Since we can't import pinyin lib, use the wp:post_name (URL-encoded slug) if available,
  // or create a date-based slug
  return '';
}

// ── Main ────────────────────────────────────────────────────────────

const xml = readFileSync(XML_PATH, 'utf-8');

// Split into items
const itemBlocks = [];
let idx = 0;
while (true) {
  const start = xml.indexOf('<item>', idx);
  if (start === -1) break;
  // Find matching </item> — items don't nest in WXR
  const end = xml.indexOf('</item>', start);
  if (end === -1) break;
  itemBlocks.push(xml.substring(start, end + 7));
  idx = end + 7;
}

console.log(`Found ${itemBlocks.length} total <item> elements`);

const articles = [];
let skippedNotPost = 0;
let skippedNotPublished = 0;

for (const block of itemBlocks) {
  const postType = cdata(tagContent(block, 'wp:post_type'));
  if (postType !== 'post') {
    skippedNotPost++;
    continue;
  }

  const status = cdata(tagContent(block, 'wp:status'));
  if (status !== 'publish') {
    skippedNotPublished++;
    continue;
  }

  const title = cdata(tagContent(block, 'title'));
  const postDate = cdata(tagContent(block, 'wp:post_date'));
  const rawContent = cdata(tagContent(block, 'content:encoded'));
  const link = tagContent(block, 'link').trim();

  // Extract categories
  const categories = [];
  const catRe = /<category domain="category"[^>]*><!\[CDATA\[(.*?)\]\]><\/category>/g;
  let catMatch;
  while ((catMatch = catRe.exec(block)) !== null) {
    categories.push(catMatch[1]);
  }

  // Extract tags
  const tags = [];
  const tagRe = /<category domain="post_tag"[^>]*><!\[CDATA\[(.*?)\]\]><\/category>/g;
  let tagMatch;
  while ((tagMatch = tagRe.exec(block)) !== null) {
    tags.push(tagMatch[1]);
  }

  // Extract wp:post_name for slug
  const wpPostName = cdata(tagContent(block, 'wp:post_name'));

  // Generate slug
  let slug = generateSlug(title);
  if (!slug && wpPostName) {
    // Decode percent-encoded Chinese slug to something usable
    try {
      const decoded = decodeURIComponent(wpPostName);
      // If decoded is Chinese, create slug from date + short hash
      if (/[\u4e00-\u9fff]/.test(decoded)) {
        const dateSlug = postDate.substring(0, 10).replace(/-/g, '');
        // Use first few chars of title transliterated or a simple hash
        slug = `${dateSlug}-${wpPostName.replace(/%/g, '').substring(0, 16).toLowerCase()}`;
      } else {
        slug = decoded.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
      }
    } catch {
      slug = wpPostName.replace(/%/g, '').substring(0, 40).toLowerCase();
    }
  }
  if (!slug) {
    slug = `post-${postDate.substring(0, 10).replace(/-/g, '')}`;
  }

  // Map categories to new system
  const mappedCategories = [...new Set(
    categories.map(c => CATEGORY_MAP[c] || 'engineering')
  )];

  // Pick primary category with priority: life > startup > industry > books > engineering
  const PRIORITY = ['life', 'startup', 'industry', 'books', 'engineering'];
  const category = PRIORITY.find(p => mappedCategories.includes(p)) || 'engineering';

  // Convert HTML to markdown
  const content = htmlToMarkdown(rawContent);

  articles.push({
    title,
    slug,
    date: postDate,
    sourceUrl: link,
    category,
    originalCategories: categories,
    tags,
    content,
  });
}

// Sort by date
articles.sort((a, b) => a.date.localeCompare(b.date));

// Write output
mkdirSync(dirname(OUT_PATH), { recursive: true });
writeFileSync(OUT_PATH, JSON.stringify(articles, null, 2), 'utf-8');

// ── Stats ───────────────────────────────────────────────────────────
console.log('\n=== WordPress XML Parse Stats ===');
console.log(`Total items in XML:       ${itemBlocks.length}`);
console.log(`Skipped (not post):       ${skippedNotPost}`);
console.log(`Skipped (not published):  ${skippedNotPublished}`);
console.log(`Published posts parsed:   ${articles.length}`);
console.log(`Output: ${OUT_PATH}`);

// Count by category
const byCat = {};
for (const a of articles) {
  byCat[a.category] = (byCat[a.category] || 0) + 1;
}
console.log('\nBy category:');
for (const [cat, count] of Object.entries(byCat).sort((a, b) => b[1] - a[1])) {
  console.log(`  ${cat}: ${count}`);
}

// Articles with missing/empty content
const empty = articles.filter(a => !a.content || a.content.length < 10);
console.log(`\nArticles with missing/empty content: ${empty.length}`);
if (empty.length > 0) {
  for (const a of empty) {
    console.log(`  - "${a.title}" (${a.date})`);
  }
}

// Show original category distribution
const byOrigCat = {};
for (const a of articles) {
  for (const c of a.originalCategories) {
    byOrigCat[c] = (byOrigCat[c] || 0) + 1;
  }
}
console.log('\nOriginal category distribution:');
for (const [cat, count] of Object.entries(byOrigCat).sort((a, b) => b[1] - a[1])) {
  console.log(`  ${cat}: ${count}`);
}

File: scripts/parse-wordpress-xml.py
#!/usr/bin/env python3
"""Parse WordPress WXR XML exports into individual raw JSON files."""

import xml.etree.ElementTree as ET
import json
import os
import re
import html
import sys
from datetime import datetime
from pathlib import Path

# WordPress XML namespaces
NS = {
    'wp': 'http://wordpress.org/export/1.2/',
    'content': 'http://purl.org/rss/1.0/modules/content/',
    'excerpt': 'http://wordpress.org/export/1.2/excerpt/',
    'dc': 'http://purl.org/dc/elements/1.1/',
}

# Category mapping from WordPress categories to 6-category system
CATEGORY_MAP = {
    # weizhiyong.com categories
    'devops': 'engineering',
    'coding': 'engineering',
    'os': 'engineering',
    'ai': 'engineering',
    'autotech': 'engineering',
    'materials': 'industry',
    'life': 'life',
    'reading': 'books',
    'book': 'books',
    'books': 'books',
    'travel': 'life',
    'family': 'life',
    'thoughts': 'writing',
    'essay': 'writing',
    'writing': 'writing',
    'startup': 'startup',
    'business': 'startup',
    'career': 'life',
    'tech': 'engineering',
    'technology': 'engineering',
    'programming': 'engineering',
    'linux': 'engineering',
    'docker': 'engineering',
    'kubernetes': 'engineering',
    'cloud': 'engineering',
    'security': 'engineering',
    'python': 'engineering',
    'javascript': 'engineering',
    'uncategorized': 'writing',  # default
}


def slugify(text):
    """Generate a URL-safe slug from text."""
    # Remove HTML tags
    text = re.sub(r'<[^>]+>', '', text)
    # Convert to lowercase
    text = text.lower().strip()
    # Replace spaces and special chars with hyphens
    text = re.sub(r'[^\w\s-]', '', text)
    text = re.sub(r'[\s_]+', '-', text)
    text = re.sub(r'-+', '-', text)
    text = text.strip('-')
    return text[:80] if text else 'untitled'


def html_to_markdown(html_content):
    """Basic HTML to Markdown conversion."""
    if not html_content:
        return ''

    text = html_content

    # Decode HTML entities
    text = html.unescape(text)

    # Headers
    for i in range(6, 0, -1):
        text = re.sub(rf'<h{i}[^>]*>(.*?)</h{i}>', r'\n' + '#' * i + r' \1\n', text, flags=re.DOTALL)

    # Bold and italic
    text = re.sub(r'<strong[^>]*>(.*?)</strong>', r'**\1**', text, flags=re.DOTALL)
    text = re.sub(r'<b[^>]*>(.*?)</b>', r'**\1**', text, flags=re.DOTALL)
    text = re.sub(r'<em[^>]*>(.*?)</em>', r'*\1*', text, flags=re.DOTALL)
    text = re.sub(r'<i[^>]*>(.*?)</i>', r'*\1*', text, flags=re.DOTALL)

    # Links
    text = re.sub(r'<a[^>]*href="([^"]*)"[^>]*>(.*?)</a>', r'[\2](\1)', text, flags=re.DOTALL)

    # Images
    text = re.sub(r'<img[^>]*src="([^"]*)"[^>]*/>', r'![](\1)', text)
    text = re.sub(r'<img[^>]*src="([^"]*)"[^>]*>', r'![](\1)', text)

    # Code blocks
    text = re.sub(r'<pre[^>]*><code[^>]*class="language-(\w+)"[^>]*>(.*?)</code></pre>',
                  r'\n```\1\n\2\n```\n', text, flags=re.DOTALL)
    text = re.sub(r'<pre[^>]*><code[^>]*>(.*?)</code></pre>',
                  r'\n```\n\1\n```\n', text, flags=re.DOTALL)
    text = re.sub(r'<pre[^>]*>(.*?)</pre>', r'\n```\n\1\n```\n', text, flags=re.DOTALL)

    # Inline code
    text = re.sub(r'<code[^>]*>(.*?)</code>', r'`\1`', text, flags=re.DOTALL)

    # Lists
    text = re.sub(r'<li[^>]*>(.*?)</li>', r'- \1', text, flags=re.DOTALL)
    text = re.sub(r'</?[ou]l[^>]*>', '', text)

    # Blockquotes
    text = re.sub(r'<blockquote[^>]*>(.*?)</blockquote>',
                  lambda m: '\n> ' + m.group(1).strip().replace('\n', '\n> ') + '\n',
                  text, flags=re.DOTALL)

    # Paragraphs and line breaks
    text = re.sub(r'<br\s*/?>', '\n', text)
    text = re.sub(r'<p[^>]*>(.*?)</p>', r'\n\1\n', text, flags=re.DOTALL)

    # Remove remaining HTML tags
    text = re.sub(r'<[^>]+>', '', text)

    # Clean up whitespace
    text = re.sub(r'\n{3,}', '\n\n', text)
    text = text.strip()

    return text


def map_category(wp_categories):
    """Map WordPress categories to the 6-category system."""
    for cat in wp_categories:
        cat_lower = cat.lower().strip()
        if cat_lower in CATEGORY_MAP:
            return CATEGORY_MAP[cat_lower]
    return 'writing'  # default


def extract_tags(wp_tags, wp_categories):
    """Extract and normalize tags."""
    tags = []
    for t in wp_tags:
        tag = t.lower().strip()
        tag = re.sub(r'[^\w-]', '-', tag)
        tag = re.sub(r'-+', '-', tag).strip('-')
        if tag and tag not in tags:
            tags.append(tag)
    # Add category keywords as tags too
    for c in wp_categories:
        c_lower = c.lower().strip()
        if c_lower not in ['uncategorized'] and c_lower not in tags:
            tags.append(c_lower)
    return tags[:8]  # max 8 tags


def parse_xml(xml_path, source_name, output_dir):
    """Parse a WordPress XML export and output raw JSON files."""
    print(f"\n=== Parsing {xml_path} (source: {source_name}) ===")

    tree = ET.parse(xml_path)
    root = tree.getroot()
    channel = root.find('channel')

    articles = []
    skipped = 0

    for item in channel.findall('item'):
        # Filter: only published posts
        post_type = item.find('wp:post_type', NS)
        post_status = item.find('wp:status', NS)

        if post_type is None or post_type.text != 'post':
            skipped += 1
            continue
        if post_status is None or post_status.text != 'publish':
            skipped += 1
            continue

        # Extract fields
        title = item.find('title').text or 'Untitled'
        content_encoded = item.find('content:encoded', NS)
        content_html = content_encoded.text if content_encoded is not None else ''

        post_date = item.find('wp:post_date', NS)
        date_str = post_date.text if post_date is not None else None

        post_name = item.find('wp:post_name', NS)
        slug = post_name.text if post_name is not None and post_name.text else slugify(title)

        # Categories and tags
        wp_categories = []
        wp_tags = []
        for cat_elem in item.findall('category'):
            domain = cat_elem.get('domain', '')
            nicename = cat_elem.get('nicename', '')
            if domain == 'category':
                wp_categories.append(nicename or cat_elem.text or '')
            elif domain == 'post_tag':
                wp_tags.append(nicename or cat_elem.text or '')

        # Convert content
        content_md = html_to_markdown(content_html)

        # Skip very short articles
        too_short = len(content_md) < 200

        # Calculate reading time (Chinese: 500 chars/min)
        char_count = len(re.sub(r'\s', '', content_md))
        reading_time = max(1, round(char_count / 500))

        # Build article record
        article = {
            'slug': slug,
            'title_zh': title,
            'source': source_name,
            'original_date': date_str,
            'category': map_category(wp_categories),
            'wp_categories': wp_categories,
            'tags': extract_tags(wp_tags, wp_categories),
            'content_zh': content_md,
            'content_html': content_html,  # keep original for reference
            'char_count': char_count,
            'reading_time_min': reading_time,
            'flags': {
                'too_short': too_short,
                'has_code_blocks': '```' in content_md,
                'has_images': '![](' in content_md,
            }
        }

        # Write individual JSON file
        out_path = os.path.join(output_dir, f'{slug}.json')
        # Handle duplicate slugs within same source
        counter = 1
        while os.path.exists(out_path):
            out_path = os.path.join(output_dir, f'{slug}-{counter}.json')
            article['slug'] = f'{slug}-{counter}'
            counter += 1

        with open(out_path, 'w', encoding='utf-8') as f:
            json.dump(article, f, ensure_ascii=False, indent=2)

        articles.append(article)

    print(f"  Parsed: {len(articles)} articles, Skipped: {skipped} non-post items")
    print(f"  Too short (<200 chars): {sum(1 for a in articles if a['flags']['too_short'])}")
    print(f"  With code: {sum(1 for a in articles if a['flags']['has_code_blocks'])}")
    print(f"  With images: {sum(1 for a in articles if a['flags']['has_images'])}")

    return articles


def main():
    base = '/workspace/extra/repos/johngai-blog/content'
    backup = '/workspace/group/blog_backup'

    # Parse weizhiyong.com
    wzy_articles = parse_xml(
        f'{backup}/WordPress.2026-03-17.xml',
        'weizhiyong',
        f'{base}/raw/weizhiyong'
    )

    # Parse johngai.com (techforge)
    jg_articles = parse_xml(
        f'{backup}/johngaistechforge.WordPress.2026-03-17.xml',
        'johngai',
        f'{base}/raw/johngai'
    )

    # Print summary
    print(f"\n=== SUMMARY ===")
    print(f"weizhiyong.com: {len(wzy_articles)} articles")
    print(f"johngai.com:    {len(jg_articles)} articles")
    print(f"Total:          {len(wzy_articles) + len(jg_articles)} articles")

    # Category distribution
    all_articles = wzy_articles + jg_articles
    cats = {}
    for a in all_articles:
        c = a['category']
        cats[c] = cats.get(c, 0) + 1
    print(f"\nCategory distribution:")
    for c, n in sorted(cats.items(), key=lambda x: -x[1]):
        print(f"  {c}: {n}")


if __name__ == '__main__':
    main()
